src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -762,6 +762,12 @@ private bool IsOpenable(bool needToUncompress, bool needToLoadIntoMemory, out st
                     message = SR.LocalFileHeaderCorrupt;
                     return false;
                 }
+                if (OffsetOfCompressedData + _uncompressedSize < _compressedSize)
```

Comment:
Can you explain why you are comparing an offset and uncompressed size to the compressed size?  Compressed/uncompressed aren't necessarily comparable.  Offset + size is not comparable to another size of the file: offset is used in above comparison to ensure the entry doesn't exceed stream length.

I would have expected the fix for this issue to compare the central directory entry length to the local header length.

Reply:
After looking through the code this is what I got:
`OffsetOfCompressedData` is represented by `_storedOffsetOfCompressedData` which is pretty much end of local header (or might better to say local header length) https://github.com/dotnet/corefx/blob/6b21df38b0d7696496bd760d6f2e0a7fe2e1b8d2/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L68-L70.
Yes compressed/uncompressed size are not comparable if not count local header file size, so idea here is local uncompressed size + local header size should be bigger than or equal to compressed size. i. e:
`if (OffsetOfCompressedData + _uncompressedSize < _compressedSize)` 
then uncompressedSize is tampered

Reply:
When we're talking about "local header length" vs "central header length" with respect to the zip file, we aren't talking about the length of the header itself, but the length of the entry as recorded in the local header.  Zip files record the length of a file in up to 3 different locations: central header, local header, or file descriptor.  Entries will only use one of the latter two (either the local header or the file descriptor).

Reply:
> When we're talking about "local header length" vs "central header length" with respect to the zip file, we aren't talking about the length of the header itself, but the length of the entry as recorded in the local header. Zip files record the length of a file in up to 3 different locations: central header, local header, or file descriptor. Entries will only use one of the latter two (either the local header or the file descriptor)

I understand that, the logic above use `local header` data of the current entry only and my comment above tried to explain how/why i used them, this might not optimal solution please suggest

> I would have expected the fix for this issue to compare the central directory entry length to the local header length.

I am not sure how to use `central directory entry length` for checking if local entry `_uncompressedSize` is tampered as we don't know how many files the zip include and from what am seeing there is not info about individual file size within the zip. I might miss something though ...

Reply:
> so idea here is local uncompressed size + local header size should be bigger than or equal to compressed size

Isn't that an arbitrary association?  Suppose you had a "bad" compression algorithm that made compressed size larger than uncompressed size (EG: the data was not very compressible and the overhead of the compression encoding made it larger).  If it was "bad enough" the compressed data size could exceed the original uncompressed size by more than the header size.  Granted, this would be unusual, but I can imagine a scenario where it is legitimate.  You could come up with some larger number X that might represent "worst case bad compression" and say if the compressed size exceeded uncompressed size by more than X (say 2 KB) you'd consider it corrupt, but that still feels pretty arbitrary.  What does the size difference matter with respect to the validity if it is self consistent?

Let me point out a few places where I think we should catch the issue (and similar ones) reported in the original issue:
1. if either of the [sizes from the central directory](https://github.com/dotnet/corefx/blob/28a926b5b38eefb18cd77b4cb0b74fdda99cdb26/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L64-L65) do not agree with the sizes in the local header when the header doesn't use data descriptor.  Currently it looks like ZipArchiveEntry isn't loading the local header on read, but [skipping it](https://github.com/dotnet/corefx/blob/28a926b5b38eefb18cd77b4cb0b74fdda99cdb26/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L754) and just decompressing the data.  Perhaps you can read it?  When reading it you may need to handle the case where the local header doesn't have size information (data descriptor bit is set).
2. when reading the data from compressed stream, make sure we don't read more than uncompressed size, here: https://github.com/dotnet/corefx/blob/28a926b5b38eefb18cd77b4cb0b74fdda99cdb26/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L358 and potentionally in Read methods of [SubReadStream](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.IO.Compression/src/System/IO/Compression/ZipCustomStreams.cs#L147)
3. in the case of a data descriptor, once we've read all the data from the stream (or if we can seek to it, check up front) and can read the descriptor, check if it matches the central directory header.

There may be more scenarios to check, but the gist is to try and ensure that all lengths are consistent and ensure we never allow reads to exceed those lengths.

Reply:
Great suggestion @ericstj thank you, totally agree with all points, let me see how i could apply your points

Reply:
One thing to keep in mind when you're looking for the size information in the local header is that there are different ways this info is stored.  I believe there are 4 cases, linking to the format we write them in below:
1. [32-bit sizes in local header](https://github.com/dotnet/corefx/blob/28a926b5b38eefb18cd77b4cb0b74fdda99cdb26/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L844-L846)
2. [64-bit sizes in local header](https://github.com/dotnet/corefx/blob/28a926b5b38eefb18cd77b4cb0b74fdda99cdb26/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L832-L840)
3. [32-bit sizes in data descriptor](https://github.com/dotnet/corefx/blob/28a926b5b38eefb18cd77b4cb0b74fdda99cdb26/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L1037-L1038)
4. [64-bit sizes in data descriptor](https://github.com/dotnet/corefx/blob/28a926b5b38eefb18cd77b4cb0b74fdda99cdb26/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L1032-L1033)

/cc @carlossanlop who recently was looking into similar issues.